File: test-errors/src/schemas/metadata/MetadataTupleType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataTupleType = void 0;
class MetadataTupleType {
    name;
    elements;
    index;
    recursive;
    nullables;
    /**
     * @internal
     */
    of_map;
    /**
     * @internal
     */
    constructor(props) {
        this.name = props.name;
        this.elements = props.elements;
        this.index = props.index;
        this.recursive = props.recursive;
        this.nullables = props.nullables;
    }
    /**
     * @internal
     */
    static _From_without_elements(props) {
        return this.create({
            name: props.name,
            index: props.index,
            elements: null,
            recursive: props.recursive,
            nullables: props.nullables.slice(),
        });
    }
    static create(props) {
        return new MetadataTupleType(props);
    }
    isRest() {
        return (this.elements.length > 0 &&
            this.elements[this.elements.length - 1].rest !== null);
    }
    toJSON() {
        return {
            name: this.name,
            index: this.index,
            elements: this.elements.map((elem) => elem.toJSON()),
            recursive: this.recursive,
            nullables: this.nullables.slice(),
        };
    }
}
exports.MetadataTupleType = MetadataTupleType;

File: packages/core/src/schemas/metadata/MetadataTupleType.go
package metadata

type MetadataTupleType struct {
	Name      string
	Elements  []*MetadataSchema
	Index     *int
	Recursive bool
	Nullables []bool

	OfMap bool
}

type MetadataTupleTypeProps struct {
	Name      string
	Elements  []*MetadataSchema
	Index     *int
	Recursive bool
	Nullables []bool
	OfMap     bool
}

func MetadataTupleTypeFromWithoutElements(props MetadataTupleTypeJSON) *MetadataTupleType {
	return CreateMetadataTupleType(MetadataTupleTypeProps{
		Name:      props.Name,
		Index:     props.Index,
		Recursive: props.Recursive,
		Nullables: props.Nullables,
	})
}

func CreateMetadataTupleType(props MetadataTupleTypeProps) *MetadataTupleType {
	return &MetadataTupleType{
		Name:      props.Name,
		Elements:  append([]*MetadataSchema(nil), props.Elements...),
		Index:     cloneIntPtr(props.Index),
		Recursive: props.Recursive,
		Nullables: append([]bool(nil), props.Nullables...),
		OfMap:     props.OfMap,
	}
}

func (m *MetadataTupleType) IsRest() bool {
	return len(m.Elements) > 0 && m.Elements[len(m.Elements)-1].Rest != nil
}

func (m *MetadataTupleType) ToJSON() MetadataTupleTypeJSON {
	elements := make([]MetadataSchemaJSON, 0, len(m.Elements))
	for _, elem := range m.Elements {
		elements = append(elements, elem.ToJSON())
	}
	return MetadataTupleTypeJSON{
		Name:      m.Name,
		Index:     cloneIntPtr(m.Index),
		Elements:  elements,
		Recursive: m.Recursive,
		Nullables: append([]bool(nil), m.Nullables...),
	}
}
